From 1dcc5484c5121e711be1b4c5ee299b2c073d75de Mon Sep 17 00:00:00 2001
From: Harry Wentland <harry.wentland@amd.com>
Date: Wed, 19 Aug 2015 15:59:50 -0400
Subject: [PATCH 388/401] drm/amdgpu: Disable DCE IRQs before suspend

Also guard disabling/enabling of interrupts before/after
S3 with the same spin lock we use elsewhere to guard
interrupt queues.

Signed-off-by: Harry Wentland <harry.wentland@amd.com>
Signed-off-by: Sanjay R Mehta <sanju.mehta@amd.com>
---
 drivers/gpu/drm/amd/amdgpu/amdgpu_dm.c     |  2 ++
 drivers/gpu/drm/amd/amdgpu/amdgpu_dm_irq.c | 33 ++++++++++++++++++++++++++++++
 drivers/gpu/drm/amd/amdgpu/amdgpu_dm_irq.h |  7 +++++++
 3 files changed, 42 insertions(+)

diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_dm.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_dm.c
index 04ffa27..7a73ba9 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_dm.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_dm.c
@@ -692,6 +692,8 @@ static int dm_suspend(void *handle)
 		DAL_ACPI_CM_POWER_STATE_D3,
 		DAL_VIDEO_POWER_SUSPEND);
 
+	amdgpu_dm_irq_suspend(adev);
+
 	return 0;
 }
 
diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_dm_irq.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_dm_irq.c
index ea55565..2e016e8 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_dm_irq.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_dm_irq.c
@@ -449,11 +449,41 @@ void amdgpu_dm_irq_fini(
 	destroy_workqueue(adev->dm.timer_workqueue);
 }
 
+int amdgpu_dm_irq_suspend(
+	struct amdgpu_device *adev)
+{
+	int src;
+	struct list_head *hnd_list_h;
+	struct list_head *hnd_list_l;
+	unsigned long irq_table_flags;
+
+	DM_IRQ_TABLE_LOCK(adev, irq_table_flags);
+
+	DRM_DEBUG_KMS("DM_IRQ: suspend\n");
+
+	/* disable HW interrupt */
+	for (src = DAL_IRQ_SOURCE_HPD1; src < DAL_IRQ_SOURCES_NUMBER; src++) {
+		hnd_list_l = &adev->dm.irq_handler_list_low_tab[src].head;
+		hnd_list_h = &adev->dm.irq_handler_list_high_tab[src];
+		if (!list_empty(hnd_list_l) || !list_empty(hnd_list_h))
+			dal_interrupt_set(adev->dm.dal, src, false);
+
+		flush_work(&adev->dm.irq_handler_list_low_tab[src].work);
+	}
+
+	DM_IRQ_TABLE_UNLOCK(adev, irq_table_flags);
+
+	return 0;
+}
+
 int amdgpu_dm_irq_resume(
 	struct amdgpu_device *adev)
 {
 	int src;
 	struct list_head *hnd_list_h, *hnd_list_l;
+	unsigned long irq_table_flags;
+
+	DM_IRQ_TABLE_LOCK(adev, irq_table_flags);
 
 	DRM_DEBUG_KMS("DM_IRQ: resume\n");
 
@@ -464,6 +494,9 @@ int amdgpu_dm_irq_resume(
 		if (!list_empty(hnd_list_l) || !list_empty(hnd_list_h))
 			dal_interrupt_set(adev->dm.dal, src, true);
 	}
+
+	DM_IRQ_TABLE_UNLOCK(adev, irq_table_flags);
+
 	return 0;
 }
 
diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_dm_irq.h b/drivers/gpu/drm/amd/amdgpu/amdgpu_dm_irq.h
index 3ca9b1a..1f3a956 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_dm_irq.h
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_dm_irq.h
@@ -106,6 +106,13 @@ void amdgpu_dm_hpd_init(struct amdgpu_device *adev);
 void amdgpu_dm_hpd_fini(struct amdgpu_device *adev);
 
 /**
+ * amdgpu_dm_irq_suspend - disable ASIC interrupt during suspend.
+ *
+ */
+int amdgpu_dm_irq_suspend(
+	struct amdgpu_device *adev);
+
+/**
  * amdgpu_dm_irq_resume - enable ASIC interrupt during resume.
  *
  */
-- 
1.9.1

